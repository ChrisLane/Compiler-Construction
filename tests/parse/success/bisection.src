// This is the iterative bisection algorithm. This should parse.
bisection (f, a, b, tol, nmax) {
    var n = 1;

    while (n <= nmax) {
        var c = a + b;
        c = c / 2;

        if (f(c) == 0 || (b - a) / 2 <= tol) {
            return $c;
        } else {};

        n = n + 1;

        if (f(c) == f(a)) {
            var a = c;
        } else {
            var b = c;
        };
    };
}
