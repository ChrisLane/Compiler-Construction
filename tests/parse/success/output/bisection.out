
bisection ( f, a, b, tol, nmax ) { 
    New ( "n" = Const 1 ) In { 
        While ( Operator ( Leq, Identifier "n" , Identifier "nmax" ) ) { 
            New ( "c" = Operator ( Plus, Identifier "a" , Identifier "b" ) ) In { 
                Asg ( Identifier "c" := Operator ( Divide, Identifier "c" , Const 2 ) ) ; 
                If ( Operator ( Or, Operator ( Equal, Application ( Identifier "f" ( Identifier "c" ) ) , Const 0 ) , Operator ( Leq, Operator ( Divide, Operator ( Minus, Identifier "b" , Identifier "a" ) , Const 2 ) , Identifier "tol" ) ) ) { 
                    Return ( Deref ( Identifier "c" ) ) 
                } Else { 
                    empty 
                } ; 
                Asg ( Identifier "n" := Operator ( Plus, Identifier "n" , Const 1 ) ) ; 
                If ( Operator ( Equal, Application ( Identifier "f" ( Identifier "c" ) ) , Application ( Identifier "f" ( Identifier "a" ) ) ) ) { 
                    New ( "a" = Identifier "c" ) In { 
                        empty 
                    } 
                } Else { 
                    New ( "b" = Identifier "c" ) In { 
                        empty 
                    } 
                } 
            } 
        } 
    } 
}

main (  ) { 
    Application ( Identifier "bisection" ( empty ) ) 
}

